## RNA-seq Session [COMPGEN2015](http://compgen2015.mdc-berlin.de/) at [BIMSB](https://www.mdc-berlin.de/40549075/de/bimsb), [MDC](https://www.mdc-berlin.de/) 

#### Date: 22 October 2015
#### Lecturer: [Mark Robinson](http://www.imls.uzh.ch/research/robinson.html), [IMLS](http://www.imls.uzh.ch/index.html), [UZH](http://www.uzh.ch/index.html)


### Rough plan 
#### 9.00-10.30: WHAT
##### 10.30-11.00: coffee
#### 11.00-12.30: HOW
##### 12.30-13.30: lunch
#### 13.30-15.00: WHY
##### 15.00-15.30: coffee
#### 15.30-17.00: DO


## WHAT
#### ice breaker
#### movo.ch survey
#### Quick intro to RNA-seq
#### Exercise in 5 groups: DE, DS, ASE, editing, de novo, exon connectivity


## HOW
#### Talk through live demo: fastest (not necessarily most sensitive) 
#### Brainstorm session: Landscape of "issues" in analyzing RNA-seq data, other pipelines


## WHY
#### Some motivation for negative binomial
#### What we know about small sample inference
#### Some recent research on methods
#### What I did in HOW versus what I would recommend


## DO
#### Students work through example (or own) dataset
#### Goal: generate (ranked) lists of DGE, DTE, DTU genes
