---
title: 'Genomic Intervals and R: Exercises'
author: "Altuna Akalin"
date: "22 Sep 2015"
output: 
  html_document:
          toc: true
          toc_depth: 1
          number_sections: true
---



# The setup

Set your working directory to the source file location in RStudio top menu:

'Session > Set Working Directory > To Source File Location'

or copy paste this your your R console in your VM.

```{r,eval=FALSE}
setwd("~/compgen2015/day6_GenomicIntervals/exercises")

```

if you have done it correctly you should see this script when you type dir()


```
> dir()
[1] "GenomicInterval.exercises.html"
```

The data for the exercises is located at `../data` folder.

Run the following to see the data files.
```
dir("../data")
```

# Operations on Genomic Intervals with GenomicRanges package

##
Create a GRanges object using the information in the table below:

| chr  | start | end  |strand | score | 
| :--- |:------| :-----| :-----|:-----|
| chr1 | 10000 | 10300 |  +    | 10 |
| chr1 | 11100 | 11500 |  -    | 20 |
| chr2 | 20000 | 20030 |  +    | 15 |


##
Import mouse (mm9 assembly) CpG islands and refseq transcripts from UCSC browser as `GRanges` objects using `rtracklayer` functions. HINT: Check the lecture material and modify the code there as necessary. 


##
Following from the exercise above, get the promoters of Refseq transcripts (-1000bp and +1000 bp of the TSS) and calculate what percentage of them overlap with CpG islands. HINT: You have to get the promoter coordinates and use `findOverlaps()` or `subsetByOverlaps()` from `GenomicRanges` package.

## 
Plot the distribution of CpG island lengths for CpG islands that overlap with the 
promoters.

##
Get canonical peaks for SP1 (peaks that are in both replicates) on chr21. Peaks for each replicate are located in `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz` and `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz` files. HINT: You need to use `findOverlaps()` or `subsetByOverlaps()` to get the subset of peaks that occur in both replicates.
EXTRA credit: use `coverage()` and slice`()` functions to get canonical peaks.

# Dealing with mapped high-throughput sequencing reads

## 
Count the reads overlapping with canonical Sp1 peaks using the BAM file for one of the replicates: `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam`. **HINT**: 
Use functions from `GenomicAlignemnts`, see [lecture notes](../lectures/genomicIntervalsTutorial.html).

# Dealing with contiguous scores over the genome

## 
Extract Views object for the promoters on chr20 from `../data/H1.ESC.H3K4me1.chr20.bw` file. Plot the first "View" as a line plot. **HINT**: see [lecture notes](../lectures/genomicIntervalsTutorial.html), adapt the code from there.

## 
Make a histogram of the maximum signal for the Views in the object you extracted above. You can use any of the view summary functions or use lapply() and write your own summary function.

## 
Get the genomic positions of maximum signal in each view and make a GRanges object. **HINT**: See ?viewRangeMaxs help page. Try to make a GRanges object out of the returned object.

# Visualizing and summarizing genomic intervals

##
Extract -500,+500 bp regions around TSSes on chr21, there are refseq files in the `../data` folder or you can
pull the data out of UCSC browser. Use SP1 ChIP-seq data 
(`../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam` ) to create an average profile of read coverage around TSSes. Following that, visualize the read coverage with a heatmap. **HINT**: All of these possible using `genomation` package functions.

##
Extract -500,+500 bp regions around TSSes on chr20. Use H3K4me3 (`../data/H1.ESC.H3K4me3.chr20.bw`) and H3K27ac (`../data/H1.ESC.H3K27ac.chr20.bw`) ChIP-seq enrichment data in the data folder and create heatmaps and average signal profiles for regions around the TSSes.

## 
Visualize one of the -500,+500 bp regions around TSS using `Gviz` functions. You should visualize both H3K4me3 and H3K27ac and the gene models.
