---
title: 'Genomic Intervals and R: Exercises'
author: "Altuna Akalin"
date: "22 Sep 2015"
output: 
  html_document:
          toc: true
          toc_depth: 1
          number_sections: true
---



# The setup

Set your working directory to the source file location in RStudio top menu:

'Session > Set Working Directory > To Source File Location'

if you have done it correctly you should see this script when you type dir()


```
> dir()
[1] "GenomicInterval.exercises.html"
```

# Operations on Genomic Intervals with GenomicRanges package

##
Create a GRanges object using the information in the table below:

| chr  | start | end  |strand | score | 
| :--- |:------| :-----| :-----|:-----|
| chr1 | 10000 | 10300 |  +    | 10 |
| chr1 | 11100 | 11500 |  -    | 20 |
| chr2 | 20000 | 10030 |  +    | 15 |


##
Import mouse (mm9 assembly) CpG islands and refseq genes from UCSC browser as GRanges objects.


##
Get promoters of refseq genes and calculate what percentage of them overlap with CpG islands.

## 
Plot the distribution of CpGi island lengths for CpG islands that overlap with the 
promoters.

##
get cannonical peaks for Sp1

# Dealing with mapped high-throughput sequencing reads

## 
Count the reads overlapping with cannonical Sp1 peaks

# Dealing with contiguous scores over the genome

